### DISCUSSION TITLE
Best way for custom CAN protocol plugin?

### DISCUSSION THREAD

**Discussion by mickeyl** [collin80/SavvyCAN#634] [General]
I'm currently designing an own CAN-bus adapter that features a custom communication protocol via BLE L2CAP. What's the best (easiest / most straightforward) way to add a CAN adapter plugin to SavvyCAN?

I already have a Swift library that implements all the nitty-gritty details of the communication, which I'd like to use. What would you recommend?

**Comment by collin80** [comment]
Well, there are two basic ways, the old way and the new way.

The new way is to directly target QT's CANBus driver interface so that any QT app could use your plugin and get CAN support. That's detailed in such places as this: https://doc.qt.io/qt-6/qcanbusdevice.html

They added that functionality WAY after SavvyCAN was started or I'd have used it from the beginning. It's probably more technically difficult to write such a plugin but I believe there are references for what QT expects of you and how to write the driver.

The old way would be to copy one of the existing "SavvyCAN only" CAN driver classes and re-implement it to use your protocol instead. This may be easier to write but also there is no documentation for how to actually do it. Very few devices have been added this way. Most everything just uses QCanBusDevice derived drivers like the socketcan driver.

**Comment by mickeyl** [comment]
Thanks. C++/Swift interoperability is a big construction site, so I will probably have to wait until Swift 5.9 has been released, but extending QtSerialBus sounds like the way to go.
